## Libraries

In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## Analyse exploratoire

In [5]:
df = pd.read_csv("vehicles.csv")

In [6]:
# Afficher les cinq premières lignes du df
df.head()

id                                                url  \
0  7222695916  https://prescott.craigslist.org/cto/d/prescott...   
1  7218891961  https://fayar.craigslist.org/ctd/d/bentonville...   
2  7221797935  https://keys.craigslist.org/cto/d/summerland-k...   
3  7222270760  https://worcester.craigslist.org/cto/d/west-br...   
4  7210384030  https://greensboro.craigslist.org/cto/d/trinit...   

                   region                         region_url  price  year  \
0                prescott    https://prescott.craigslist.org   6000   NaN   
1            fayetteville       https://fayar.craigslist.org  11900   NaN   
2            florida keys        https://keys.craigslist.org  21000   NaN   
3  worcester / central MA   https://worcester.craigslist.org   1500   NaN   
4              greensboro  https://greensboro.craigslist.org   4900   NaN   

  manufacturer model condition cylinders  ... size  type paint_color  \
0          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   
1          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   
2          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   
3          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   
4          NaN   NaN       NaN       NaN  ...  NaN   NaN         NaN   

  image_url description county state lat long posting_date  
0       NaN         NaN    NaN    az NaN  NaN          NaN  
1       NaN         NaN    NaN    ar NaN  NaN          NaN  
2       NaN         NaN    NaN    fl NaN  NaN          NaN  
3       NaN         NaN    NaN    ma NaN  NaN          NaN  
4       NaN         NaN    NaN    nc NaN  NaN          NaN  

[5 rows x 26 columns]

In [7]:
# Afficher les cinq dernières lignes du df
df.tail()

id                                                url  \
426875  7301591192  https://wyoming.craigslist.org/ctd/d/atlanta-2...   
426876  7301591187  https://wyoming.craigslist.org/ctd/d/atlanta-2...   
426877  7301591147  https://wyoming.craigslist.org/ctd/d/atlanta-2...   
426878  7301591140  https://wyoming.craigslist.org/ctd/d/atlanta-2...   
426879  7301591129  https://wyoming.craigslist.org/ctd/d/atlanta-2...   

         region                      region_url  price    year manufacturer  \
426875  wyoming  https://wyoming.craigslist.org  23590  2019.0       nissan   
426876  wyoming  https://wyoming.craigslist.org  30590  2020.0        volvo   
426877  wyoming  https://wyoming.craigslist.org  34990  2020.0     cadillac   
426878  wyoming  https://wyoming.craigslist.org  28990  2018.0        lexus   
426879  wyoming  https://wyoming.craigslist.org  30590  2019.0          bmw   

                           model condition    cylinders  ... size       type  \
426875         maxima s sedan 4d      good  6 cylinders  ...  NaN      sedan   
426876  s60 t5 momentum sedan 4d      good          NaN  ...  NaN      sedan   
426877          xt4 sport suv 4d      good          NaN  ...  NaN  hatchback   
426878           es 350 sedan 4d      good  6 cylinders  ...  NaN      sedan   
426879  4 series 430i gran coupe      good          NaN  ...  NaN      coupe   

       paint_color                                          image_url  \
426875         NaN  https://images.craigslist.org/00o0o_iiraFnHg8q...   
426876         red  https://images.craigslist.org/00x0x_15sbgnxCIS...   
426877       white  https://images.craigslist.org/00L0L_farM7bxnxR...   
426878      silver  https://images.craigslist.org/00z0z_bKnIVGLkDT...   
426879         NaN  https://images.craigslist.org/00Y0Y_lEUocjyRxa...   

                                              description county state  \
426875  Carvana is the safer way to buy a car During t...    NaN    wy   
426876  Carvana is the safer way to buy a car During t...    NaN    wy   
426877  Carvana is the safer way to buy a car During t...    NaN    wy   
426878  Carvana is the safer way to buy a car During t...    NaN    wy   
426879  Carvana is the safer way to buy a car During t...    NaN    wy   

              lat       long              posting_date  
426875  33.786500 -84.445400  2021-04-04T03:21:31-0600  
426876  33.786500 -84.445400  2021-04-04T03:21:29-0600  
426877  33.779214 -84.411811  2021-04-04T03:21:17-0600  
426878  33.786500 -84.445400  2021-04-04T03:21:11-0600  
426879  33.779214 -84.411811  2021-04-04T03:21:07-0600  

[5 rows x 26 columns]

In [8]:
# Affciher les lignes du df, aléatoirement
df.sample(n=10)

id                                                url  \
159928  7311771881  https://masoncity.craigslist.org/ctd/d/atlanta...   
378677  7313359355  https://mcallen.craigslist.org/cto/d/alamo-for...   
319334  7314979287  https://medford.craigslist.org/cto/d/ashland-1...   
182434  7310797642  https://baltimore.craigslist.org/ctd/d/glen-bu...   
80141   7306928908  https://fortcollins.craigslist.org/cto/d/lovel...   
59917   7314059609  https://santamaria.craigslist.org/ctd/d/los-an...   
87426   7304223643  https://newhaven.craigslist.org/ctd/d/new-have...   
81196   7311592765  https://pueblo.craigslist.org/ctd/d/pueblo-201...   
377547  7316191168  https://laredo.craigslist.org/ctd/d/liberty-hi...   
126888  7309160947  https://savannah.craigslist.org/ctd/d/savannah...   

                         region                          region_url  price  \
159928               mason city    https://masoncity.craigslist.org  29590   
378677       mcallen / edinburg      https://mcallen.craigslist.org      1   
319334          medford-ashland      https://medford.craigslist.org  18000   
182434                baltimore    https://baltimore.craigslist.org  25765   
80141   fort collins / north CO  https://fortcollins.craigslist.org  18500   
59917               santa maria   https://santamaria.craigslist.org  35000   
87426                 new haven     https://newhaven.craigslist.org  25990   
81196                    pueblo       https://pueblo.craigslist.org   8500   
377547                   laredo       https://laredo.craigslist.org  46995   
126888    savannah / hinesville     https://savannah.craigslist.org  31590   

          year   manufacturer                   model  condition    cylinders  \
159928  2017.0           ford    expedition xlt sport       good  6 cylinders   
378677  1997.0           ford                f250 xlt        NaN          NaN   
319334  1972.0      chevrolet                    c-10        NaN          NaN   
182434  2014.0           jeep      wrangler unlimited        NaN          NaN   
80141   1981.0           fiat      pininfarina spider   like new  4 cylinders   
59917   2015.0  mercedes-benz  benz sprinter 2500 ext  excellent  4 cylinders   
87426   2016.0         toyota   tacoma access cab sr5       good          NaN   
81196   2013.0           ford                edge sel       good  6 cylinders   
377547  2014.0           ford       f250 platinum 4x4        NaN          NaN   
126888  2019.0       cadillac        xt4 sport suv 4d       good          NaN   

        ...     size         type paint_color  \
159928  ...      NaN          SUV        blue   
378677  ...      NaN        truck         NaN   
319334  ...      NaN          NaN         NaN   
182434  ...      NaN          NaN         NaN   
80141   ...  compact  convertible      custom   
59917   ...      NaN          van       white   
87426   ...      NaN       pickup      silver   
81196   ...      NaN          NaN         NaN   
377547  ...      NaN          NaN         NaN   
126888  ...      NaN    hatchback       black   

                                                image_url  \
159928  https://images.craigslist.org/00N0N_1xMPvfxRAI...   
378677  https://images.craigslist.org/00A0A_ZaSvhNnSYc...   
319334  https://images.craigslist.org/00s0s_4jBuWdML03...   
182434  https://images.craigslist.org/00C0C_6PwXFFXGgL...   
80141   https://images.craigslist.org/00U0U_1ugz3GuRWq...   
59917   https://images.craigslist.org/00m0m_1vuCRDj0B4...   
87426   https://images.craigslist.org/00w0w_bJYQv4CFjd...   
81196   https://images.craigslist.org/00M0M_jEN1EkZSqo...   
377547  https://images.craigslist.org/00808_hxHMwjOzuF...   
126888  https://images.craigslist.org/00v0v_Nmw3hmqa7i...   

                                              description county state  \
159928  Carvana is the safer way to buy a car During t...    NaN    ia   
378677  1997 Ford F-250 powerstroke diesel for sale 4 ...    NaN    tx   
319334  one owner c-10 shor

In [9]:
# Dimension du df c-a-d le nombre de lignes et colonnes
df.shape

(426880, 26)

In [10]:
# Nombre de colonnes dans df
df.columns

Index(['id', 'url', 'region', 'region_url', 'price', 'year', 'manufacturer',
       'model', 'condition', 'cylinders', 'fuel', 'odometer', 'title_status',
       'transmission', 'VIN', 'drive', 'size', 'type', 'paint_color',
       'image_url', 'description', 'county', 'state', 'lat', 'long',
       'posting_date'],
      dtype='object')

In [11]:
# Le type de données de chaque colonne du df
df.dtypes

id                int64
url              object
region           object
region_url       object
price             int64
year            float64
manufacturer     object
model            object
condition        object
cylinders        object
fuel             object
odometer        float64
title_status     object
transmission     object
VIN              object
drive            object
size             object
type             object
paint_color      object
image_url        object
description      object
county          float64
state            object
lat             float64
long            float64
posting_date     object
dtype: object

La commande ci-dessous, nous me permet d'avoir des informations sur le dataset comme le type de données et le nombre total des valeurs non nulles dans chaque colonne.

Cela nous permet didentifier les valeurs manquantes et les types de données incorrectes 

In [12]:

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 426880 entries, 0 to 426879
Data columns (total 26 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   id            426880 non-null  int64  
 1   url           426880 non-null  object 
 2   region        426880 non-null  object 
 3   region_url    426880 non-null  object 
 4   price         426880 non-null  int64  
 5   year          425675 non-null  float64
 6   manufacturer  409234 non-null  object 
 7   model         421603 non-null  object 
 8   condition     252776 non-null  object 
 9   cylinders     249202 non-null  object 
 10  fuel          423867 non-null  object 
 11  odometer      422480 non-null  float64
 12  title_status  418638 non-null  object 
 13  transmission  424324 non-null  object 
 14  VIN           265838 non-null  object 
 15  drive         296313 non-null  object 
 16  size          120519 non-null  object 
 17  type          334022 non-null  object 
 18  pain

In [13]:
#Statistiques descriptives pour les colonnes numériques (moyenne, écart-type, min, max, quartiles)
df.describe()

id         price           year      odometer  county  \
count  4.268800e+05  4.268800e+05  425675.000000  4.224800e+05     0.0   
mean   7.311487e+09  7.519903e+04    2011.235191  9.804333e+04     NaN   
std    4.473170e+06  1.218228e+07       9.452120  2.138815e+05     NaN   
min    7.207408e+09  0.000000e+00    1900.000000  0.000000e+00     NaN   
25%    7.308143e+09  5.900000e+03    2008.000000  3.770400e+04     NaN   
50%    7.312621e+09  1.395000e+04    2013.000000  8.554800e+04     NaN   
75%    7.315254e+09  2.648575e+04    2017.000000  1.335425e+05     NaN   
max    7.317101e+09  3.736929e+09    2022.000000  1.000000e+07     NaN   

                 lat           long  
count  420331.000000  420331.000000  
mean       38.493940     -94.748599  
std         5.841533      18.365462  
min       -84.122245    -159.827728  
25%        34.601900    -111.939847  
50%        39.150100     -88.432600  
75%        42.398900     -80.832039  
max        82.390818     173.885502

In [14]:
# Statistiques descriptives pour des colonnes catégorielles (nombre de valeurs uniques, fréquence des plus courantes).
df.describe(include='object')

url    region  \
count                                              426880    426880   
unique                                             426880       404   
top     https://wyoming.craigslist.org/ctd/d/atlanta-2...  columbus   
freq                                                    1      3608   

                            region_url manufacturer   model condition  \
count                           426880       409234  421603    252776   
unique                             413           42   29667         6   
top     https://spokane.craigslist.org         ford   f-150      good   
freq                              2988        70985    8009    121456   

          cylinders    fuel title_status transmission                VIN  \
count        249202  423867       418638       424324             265838   
unique            8       5            6            3             118264   
top     6 cylinders     gas        clean    automatic  1FMJU1JT1HEA52352   
freq          94169  356209       405117       336524                261   

         drive       size    type paint_color  \
count   296313     120519  334022      296677   
unique       3          4      13          12   
top        4wd  full-size   sedan       white   
freq    131904      63465   87056       79285   

                                                image_url  \
count                                              426812   
unique                                             241899   
top     https://images.craigslist.org/00N0N_1xMPvfxRAI...   
freq                                                 7357   

                                              description   state  \
count                                              426810  426880   
unique                                             360911      51   
top     35 VEHICLES PRICED UNDER $3000!!!  BIG TIME! T...      ca   
freq                                                  231   50614   

                    posting_date  
count                     426812  
unique                    381536  
top     2021-04-23T22:13:05-0400  
freq                          12

## Nettoyage

Cette partie consiste à manipuler le dataset dans le but de le nettoyer et de le traiter.

Pour ce faire, nous allons procéder en gérant les valeurs manquantes, doublons, abbérantes et corriger si nécessaire les types de données des colonnes non conformes



### Gestion des valeurs manquantes

In [15]:
# Verifier la présence des valeurs manaquantes

valeurs_manquantes = df.isnull().sum()
print(valeurs_manquantes)

id                   0
url                  0
region               0
region_url           0
price                0
year              1205
manufacturer     17646
model             5277
condition       174104
cylinders       177678
fuel              3013
odometer          4400
title_status      8242
transmission      2556
VIN             161042
drive           130567
size            306361
type             92858
paint_color     130203
image_url           68
description         70
county          426880
state                0
lat               6549
long              6549
posting_date        68
dtype: int64



#### Selectionner les variables non pertinentes 

Ce son celles qui ne sont assez décisif pour acheter une voiture puis à les retirer du dataset.

Ces variables sont :

*County* qui n'a aucunes données, spécifie à renseigner le lieu où l'annonce a été postée.

*Size* qui contient beaucoup de valeurs manquantes d'ordre 300k, est une variable qu'on peut s'empasser lors des analyses car la variable *Type* donne l'information sur le design du véhicule.

*VIN* présente beaucoup de NA d'ordre 100k, n'est nécessaire pour l'analyse car elle symbolise le matricule de la voiture. Le matricule d'une voiture est une donnée qui peut être changeable par son nouveau propriétaire.

*Description*  n'a pas de pertinence car elle est similaire à page 'à propos' d'un site. Elle n'apporte pas d'information cruciale sur le véhicule.

*lat & long* represente les coordonnées géographiques de l'endroit où l'annonce a été faite.

*image_url* non nécessaire pour l'analyse, aussi les liens sont non fonctionnels qui peut être du au fait que l'annonce ait été retiré par l'achat du vehicule ou pas.

Enfin les variables *id, url, region, et region_url* certe n'ont aucune valeurs manquantes cependant elles servent à titre d'identifier ou de référence de l'annonce.




In [17]:
colonnes_a_supprimer = df[['VIN', 'description', 'county', 'id','lat', 'long', 
                           'size', 'url', 'region', 'region_url', 'image_url']]

In [18]:
colonnes_a_supprimer

VIN                                        description  \
0                     NaN                                                NaN   
1                     NaN                                                NaN   
2                     NaN                                                NaN   
3                     NaN                                                NaN   
4                     NaN                                                NaN   
...                   ...                                                ...   
426875  1N4AA6AV6KC367801  Carvana is the safer way to buy a car During t...   
426876  7JR102FKXLG042696  Carvana is the safer way to buy a car During t...   
426877  1GYFZFR46LF088296  Carvana is the safer way to buy a car During t...   
426878  58ABK1GG4JU103853  Carvana is the safer way to buy a car During t...   
426879  WBA4J1C58KBM14708  Carvana is the safer way to buy a car During t...   

        county          id        lat       long size  \
0          NaN  7222695916        NaN        NaN  NaN   
1          NaN  7218891961        NaN        NaN  NaN   
2          NaN  7221797935        NaN        NaN  NaN   
3          NaN  7222270760        NaN        NaN  NaN   
4          NaN  7210384030        NaN        NaN  NaN   
...        ...         ...        ...        ...  ...   
426875     NaN  7301591192  33.786500 -84.445400  NaN   
426876     NaN  7301591187  33.786500 -84.445400  NaN   
426877     NaN  7301591147  33.779214 -84.411811  NaN   
426878     NaN  7301591140  33.786500 -84.445400  NaN   
426879     NaN  7301591129  33.779214 -84.411811  NaN   

                                                      url  \
0       https://prescott.craigslist.org/cto/d/prescott...   
1       https://fayar.craigslist.org/ctd/d/bentonville...   
2       https://keys.craigslist.org/cto/d/summerland-k...   
3       https://worcester.craigslist.org/cto/d/west-br...   
4       https://greensboro.craigslist.org/cto/d/trinit...   
...                                                   ...   
426875  https://wyoming.craigslist.org/ctd/d/atlanta-2...   
426876  https://wyoming.craigslist.org/ctd/d/atlanta-2...   
426877  https://wyoming.craigslist.org/ctd/d/atlanta-2...   
426878  https://wyoming.craigslist.org/ctd/d/atlanta-2...   
426879  https://wyoming.craigslist.org/ctd/d/atlanta-2...   

                        region                         region_url  \
0                     prescott    https://prescott.craigslist.org   
1                 fayetteville       https://fayar.craigslist.org   
2                 florida keys        https://keys.craigslist.org   
3       worcester / central MA   https://worcester.craigslist.org   
4                   greensboro  https://greensboro.craigslist.org   
...                        ...                                ...   
426875                 wyoming     https://wyoming.craigslist.org   
426876                 wyoming     https://wyoming.craigslist.org   
426877                 wyoming     https://wyoming.craigslist.org   
426878                 wyoming     https://wyoming.craigslist.org   
426879                 wyoming     https://wyoming.craigslist.org   

                                                image_url  
0                                                     NaN  
1                                                     NaN  
2                                                     NaN  
3                                                     NaN  
4                                                     NaN  
...                                                   ...  
426875  https://images.craigslist.org/00o0o_iiraFnHg8q...  
426876  https://images.craigslist.org/00x0x_15sbgnxCIS...  
426877  https://images.craigslist.org/00L0L_farM7bxnxR...  
426878  https://images.craigslist.org/00z0z_bKnIVGLkDT...  
426879  https://images.craigslist.org/00Y0Y_lEUocjyRxa...  

[426880 rows x 11 columns]

In [19]:
df.drop(colonnes_a_supprimer, axis=1, inplace=True)

In [ ]:
df

price    year manufacturer                     model condition  \
0        6000     NaN          NaN                       NaN       NaN   
1       11900     NaN          NaN                       NaN       NaN   
2       21000     NaN          NaN                       NaN       NaN   
3        1500     NaN          NaN                       NaN       NaN   
4        4900     NaN          NaN                       NaN       NaN   
...       ...     ...          ...                       ...       ...   
426875  23590  2019.0       nissan         maxima s sedan 4d      good   
426876  30590  2020.0        volvo  s60 t5 momentum sedan 4d      good   
426877  34990  2020.0     cadillac          xt4 sport suv 4d      good   
426878  28990  2018.0        lexus           es 350 sedan 4d      good   
426879  30590  2019.0          bmw  4 series 430i gran coupe      good   

          cylinders    fuel  odometer title_status transmission drive  \
0               NaN     NaN       NaN          NaN          NaN   NaN   
1               NaN     NaN       NaN          NaN          NaN   NaN   
2               NaN     NaN       NaN          NaN          NaN   NaN   
3               NaN     NaN       NaN          NaN          NaN   NaN   
4               NaN     NaN       NaN          NaN          NaN   NaN   
...             ...     ...       ...          ...          ...   ...   
426875  6 cylinders     gas   32226.0        clean        other   fwd   
426876          NaN     gas   12029.0        clean        other   fwd   
426877          NaN  diesel    4174.0        clean        other   NaN   
426878  6 cylinders     gas   30112.0        clean        other   fwd   
426879          NaN     gas   22716.0        clean        other   rwd   

             type paint_color state              posting_date  
0             NaN         NaN    az                       NaN  
1             NaN         NaN    ar                       NaN  
2             NaN         NaN    fl                       NaN  
3             NaN         NaN    ma                       NaN  
4             NaN         NaN    nc                       NaN  
...           ...         ...   ...                       ...  
426875      sedan         NaN    wy  2021-04-04T03:21:31-0600  
426876      sedan         red    wy  2021-04-04T03:21:29-0600  
426877  hatchback       white    wy  2021-04-04T03:21:17-0600  
426878      sedan      silver    wy  2021-04-04T03:21:11-0600  
426879      coupe         NaN    wy  2021-04-04T03:21:07-0600  

[426880 rows x 15 columns]

#### Etudier  des imputations dans les varaibles 
##### Analyse des variables title_status et condition

Les variables title_status et condition donnent la même information sur l'état du véhicule. Cependant les valeurs de la colonne *title_status* sont plus réalistes que l'autre colonne.

En effet, *title_status* ces observations sont objectives tandis que celles de *condition* sont subjectives c'est à dire basées sur le point de vue du vendeur.  

In [20]:
df['title_status'].unique()

array([nan, 'clean', 'rebuilt', 'lien', 'salvage', 'missing',
       'parts only'], dtype=object)

In [ ]:
df['condition'].unique()

array([nan, 'good', 'excellent', 'fair', 'like new', 'new', 'salvage'],
      dtype=object)

*Qu'est-ce qu'une Matrice de Contingence ?*

Une matrice de contingence (ou tableau croisé dynamique) est un tableau qui affiche la distribution conjointe de deux variables catégorielles (ou qualitatives). 

En d'autres termes, elle permet de comprendre comment les fréquences d'une catégorie d'une variable se répartissent selon les catégories d'une autre variable.

In [ ]:
print("\nMatrice de contingence entre title_status et condition:")
contingency_table = pd.crosstab(df['title_status'], df['condition'])
print(contingency_table)


Matrice de contingence entre title_status et condition:
condition     excellent  fair    good  like new   new  salvage
title_status                                                  
clean             91734  6156  118461     19870  1226      225
lien                586    47     389       265    25        3
missing              57   188     208         8     5       49
parts only           18    27      29         9     4       41
rebuilt            2823   142    1362       791    30       23
salvage             880   209     893       235    15      260


Le constat montre qu'il y a de forte corrélation, ce qui permet de faire de l'imputation conditionnelle afin de compléter les valeurs NAN de la colonne *condition*

Par exemple, nous constatons que:
- 91734 voitures sont à la fois *clean et excellent*
- 118 461 voitures sont à la fois "clean et good"

La plupart des voitures sont en *excellent, good* soit en statut *clean* au lieu de *salvage*




In [ ]:
print("\nMatrice de contingence entre title_status et condition avec NaN:")
contingency_table = pd.crosstab(df['title_status'], df['condition'], dropna=False)
print(contingency_table)


Matrice de contingence entre title_status et condition avec NaN:
condition     excellent  fair    good  like new   new  salvage     NaN
title_status                                                          
clean             91734  6156  118461     19870  1226      225  167445
lien                586    47     389       265    25        3     107
missing              57   188     208         8     5       49     299
parts only           18    27      29         9     4       41      70
rebuilt            2823   142    1362       791    30       23    2048
salvage             880   209     893       235    15      260    1376
NaN                5369     0     114         0     0        0    2759


Etant donnée que ces deux variables sont corrélées, on peut utiliser l'approche d'imputer les valeurs d'une variable en fonction des valeurs de l'autre variable. Cela s'appelle *l'imputation conditionnelle*

In [21]:
# Imputer les valeurs NA de title_status par missing 
# qui renvoit unknown (non-renseigné)

df['title_status'] = df['title_status'].fillna('missing')

In [22]:
#vérifier
df['title_status'].isna().sum()

np.int64(0)

**Explication du code**

Dans la logique, un véhicule présentant un état de statut 'salvage ou endommagé', nous permet de présumer que sa condition sera aussi en mauvaise état qui correspond dans la variable *condition* à 'fear' ou 'salvage'


On selectionne les lignes qui ont d'une part une valeur **'NA'** dans la colonne *condition* et d'autre part une valeur **'salvage'** dans la colonne *title_status*

Dans l'affichage, seule la colonne **condition** appraitra vu qu'elle est renseigné comme deuxième argument de la location (loc) qui sera affecté par les valeurs *salvage* dans les champs NA.

De même avec une valeur *part_only* de la colonne *title_status* va plus correspondre à une condition *salvage ou fear*

In [23]:

df.loc[(df['condition'].isnull()) & (df['title_status'] == 'salvage'), 'condition'] = 'salvage'


In [24]:
df.loc[(df['condition'].isnull()) & (df['title_status'] == 'parts only'), 'condition'] = 'salvage' # ou 'fair'

In [ ]:
print("\nMatrice de contingence entre title_status et condition avec NaN après traitement:")
contingency_table = pd.crosstab(df['title_status'], df['condition'], dropna=False)
print(contingency_table)


Matrice de contingence entre title_status et condition avec NaN:
condition     excellent  fair    good  like new   new  salvage     NaN
title_status                                                          
clean             91734  6156  118461     19870  1226      225  167445
lien                586    47     389       265    25        3     107
missing            5426   188     322         8     5       49    3058
parts only           18    27      29         9     4      111       0
rebuilt            2823   142    1362       791    30       23    2048
salvage             880   209     893       235    15     1636       0


In [25]:
#la méthode .mode() prend la valeur la plus fréquente
df['condition'] = df.groupby('title_status')['condition'].transform(
    lambda x: x.fillna(x.mode()[0] 
                       if not x.mode().empty else 'unknown condition'))


In [26]:
print("\nMatrice de contingence entre title_status et condition avec NaN après traitement:")
contingency_table = pd.crosstab(df['title_status'], df['condition'], dropna=False)
print(contingency_table)


Matrice de contingence entre title_status et condition avec NaN après traitement:
condition     excellent  fair    good  like new   new  salvage
title_status                                                  
clean             91734  6156  285906     19870  1226      225
lien                693    47     389       265    25        3
missing            8484   188     322         8     5       49
parts only           18    27      29         9     4      111
rebuilt            4871   142    1362       791    30       23
salvage             880   209     893       235    15     1636


In [30]:
valeurs_manquantes = df.isnull().sum()
print(valeurs_manquantes)

price                0
year              1205
manufacturer     17646
model             5277
condition            0
cylinders       177678
fuel              3013
odometer          4400
title_status         0
transmission      2556
drive           130567
type             92858
paint_color     130203
state                0
posting_date        68
dtype: int64


In [27]:
df.sort_values(by='odometer')

price    year manufacturer               model  condition  \
155630      0  2006.0         ford                f550       good   
56421    9900  2015.0     chrysler                 200  excellent   
205001      0  2014.0         ford            f550 4x4       good   
166284   4499  1996.0          NaN     Sea Ray Sea Ray       good   
56420   10900  2010.0        buick             enclave  excellent   
...       ...     ...          ...                 ...        ...   
421748      0  2014.0   volkswagen              passat       good   
421780      0  2021.0    chevrolet  5500 4 x 4 chassis        new   
423072   1400  2003.0     cadillac             deville       good   
423127   4500  2011.0        buick             enclave       good   
424638   3900  2004.0     infiniti                fx35       good   

          cylinders    fuel  odometer title_status transmission drive   type  \
155630          NaN  diesel       0.0        clean        other   NaN  other   
56421   4 cylinders     gas       0.0        clean    automatic   fwd  sedan   
205001          NaN  diesel       0.0        clean        other   NaN  other   
166284          NaN     gas       0.0        clean        other   NaN  other   
56420   6 cylinders     gas       0.0        clean    automatic   4wd  wagon   
...             ...     ...       ...          ...          ...   ...    ...   
421748  4 cylinders     gas       NaN        clean    automatic   fwd  sedan   
421780          NaN     gas       NaN        clean    automatic   NaN    NaN   
423072          NaN     gas       NaN        clean    automatic   NaN  sedan   
423127          NaN     gas       NaN        clean    automatic   fwd    SUV   
424638          NaN     gas       NaN        clean    automatic   4wd  wagon   

       paint_color state              posting_date  
155630       white    ia  2021-04-15T16:40:01-0500  
56421         grey    ca  2021-04-30T22:12:29-0700  
205001       white    mi  2021-04-12T15:14:02-0400  
166284       white    ks  2021-04-25T17:01:34-0500  
56420       custom    ca  2021-04-30T22:12:32-0700  
...            ...   ...                       ...  
421748      silver    wi  2021-04-08T10:30:19-0500  
421780         NaN    wi  2021-04-08T07:45:27-0500  
423072       white    wi  2021-04-29T13:20:38-0500  
423127       black    wi  2021-04-29T08:30:28-0500  
424638         NaN    wi  2021-04-17T14:21:47-0500  

[426880 rows x 15 columns]

On constate qu'il y'a des kilométrages qui sont nulles sachant que nous faisons face à des voitures d'occasion. Dans la logique, les voitures doivent avoir un kilométrage supérieur à 0.

Etant donnée que ces lignes sont minimes dans le dataset alors on peut les retirer car elles peuvent induire en erreur par le fait que ces informations ne sont pas adéquates. De plus leur prix sont inférieurs à 0.

In [31]:
requete = df['odometer'] == 0
df[requete]

price    year manufacturer             model  condition    cylinders  \
538         0  2012.0    chevrolet  silverado 2500hd       good  8 cylinders   
563         0  2006.0         ford              f550       good          NaN   
694      3980  2013.0          kia               rio  excellent  4 cylinders   
889         0  2006.0         jeep           liberty       good  6 cylinders   
985         0  2011.0    chevrolet             cruze       good  4 cylinders   
...       ...     ...          ...               ...        ...          ...   
425904      0  2014.0          ram               NaN       good          NaN   
425905      0  2014.0          ram               NaN       good          NaN   
426002      0  2014.0          ram               NaN       good          NaN   
426518      0  2014.0          ram               NaN       good          NaN   
426638      0  2015.0        dodge            intern       good          NaN   

          fuel  odometer title_status transmission drive   type paint_color  \
538        NaN       0.0        clean    automatic   rwd    NaN       white   
563     diesel       0.0        clean        other   NaN  other       white   
694        gas       0.0      missing       manual   fwd  sedan       white   
889        gas       0.0        clean    automatic   4wd    SUV         red   
985        gas       0.0        clean    automatic   rwd  sedan         red   
...        ...       ...          ...          ...   ...    ...         ...   
425904  diesel       0.0        clean    automatic   NaN  other       white   
425905  diesel       0.0        clean    automatic   NaN  other       white   
426002  diesel       0.0        clean    automatic   NaN  other       white   
426518  diesel       0.0        clean    automatic   NaN  other       white   
426638  diesel       0.0        clean        other   NaN  other       white   

       state              posting_date  
538       al  2021-04-27T19:50:22-0500  
563       al  2021-04-27T14:10:03-0500  
694       al  2021-04-24T13:50:11-0500  
889       al  2021-04-22T17:40:42-0500  
985       al  2021-04-20T13:20:43-0500  
...      ...                       ...  
425904    wi  2021-04-20T19:33:56-0500  
425905    wi  2021-04-20T19:30:08-0500  
426002    wi  2021-04-16T02:00:05-0500  
426518    wy  2021-04-21T22:45:37-0600  
426638    wy  2021-04-15T10:40:07-0600  

[1965 rows x 15 columns]

<bound method DataFrame.count of         price    year manufacturer             model  condition    cylinders  \
538         0  2012.0    chevrolet  silverado 2500hd       good  8 cylinders   
563         0  2006.0         ford              f550       good          NaN   
694      3980  2013.0          kia               rio  excellent  4 cylinders   
889         0  2006.0         jeep           liberty       good  6 cylinders   
985         0  2011.0    chevrolet             cruze       good  4 cylinders   
...       ...     ...          ...               ...        ...          ...   
425904      0  2014.0          ram               NaN       good          NaN   
425905      0  2014.0          ram               NaN       good          NaN   
426002      0  2014.0          ram               NaN       good          NaN   
426518      0  2014.0          ram               NaN       good          NaN   
426638      0  2015.0        dodge            intern       good          NaN   

      

### Correction des types de données

### Gestion des doublons

### Gestion des abérrants (outlet)

## Axes d'analyse

## Vectorisation